cli/src/commands/duplicate.rs

```diff
@@ -57,37 +87,164 @@ pub(crate) fn cmd_duplicate(
+    } else if !args.insert_after.is_empty() {
+        let parent_commits = workspace_command
+            .resolve_some_revsets_default_single(ui, &args.insert_after)?
+            .into_iter()
+            .collect_vec();
+        parent_commit_ids = parent_commits.iter().ids().cloned().collect();
+        let parents_expression = RevsetExpression::commits(parent_commit_ids.clone());
+        let children_expression = parents_expression.children();
+        children_commit_ids = children_expression
+            .clone()
+            .evaluate(workspace_command.repo().as_ref())
+            .map_err(|err| err.expect_backend_error())?
             .iter()
-            .map(|id| duplicated_old_to_new.get(id).map_or(id, |c| c.id()).clone())
-            .collect();
-        let new_commit = mut_repo
-            .rewrite_commit(command.settings(), &original_commit)
-            .generate_new_change_id()
-            .set_parents(new_parents)
-            .write()?;
-        duplicated_old_to_new.insert(original_commit_id, new_commit);
+            .try_collect()?;
+        workspace_command.check_rewritable(&children_commit_ids)?;
+        ensure_no_commit_loop(
+            workspace_command.repo(),
+            &children_expression,
+            &parents_expression,
+        )?;
+    } else if !args.destination.is_empty() {
+        let parent_commits = workspace_command
+            .resolve_some_revsets_default_single(ui, &args.destination)?
+            .into_iter()
+            .collect_vec();
+        parent_commit_ids = parent_commits.iter().ids().cloned().collect();
+        children_commit_ids = vec![];
+    } else {
+        parent_commit_ids = vec![];
+        children_commit_ids = vec![];
+    };
+
+    let mut tx = workspace_command.start_transaction();
+
+    if !parent_commit_ids.is_empty() {
+        for commit_id in &to_duplicate {
+            for parent_commit_id in &parent_commit_ids {
+                if tx.repo().index().is_ancestor(commit_id, parent_commit_id) {
+                    writeln!(
+                        ui.warning_default(),
+                        "Duplicating commit {} as a descendant of itself",
+                        short_commit_hash(commit_id)
+                    )?;
+                    break;
+                }
+            }
+        }
+
+        for commit_id in &to_duplicate {
+            for child_commit_id in &children_commit_ids {
+                if tx.repo().index().is_ancestor(child_commit_id, commit_id) {
+                    writeln!(
+                        ui.warning_default(),
+                        "Duplicating commit {} as an ancestor of itself",
+                        short_commit_hash(commit_id)
+                    )?;
+                    break;
+                }
+            }
+        }
     }
 
+    let num_to_duplicate = to_duplicate.len();
+    let DuplicateCommitsResult {
+        duplicated_commits,
+        num_rebased,
+    } = if parent_commit_ids.is_empty() {
```

Comment:
nit: might be better to test `args`, not the computed commit_ids. It's not so obvious that `parent_commit_ids` can never be empty if any of `-d`/`-A`/`-B` are specified.